src/gp_symbolic_regression.py

```diff
@@ -0,0 +1,162 @@
+"""
+A genetic programming for symbolic regression.
+
+Data from some unknown function is read in from a file. Rows represent observations of the function and columns
+represent the variables. The last column in the file must be the dependant variable. The file can contain arbitrarily
+many variables but should contain at least two --- one independent and one dependent variable.
+"""
+
+import operator
+import os.path
+from functools import partial
+from random import randint
+
+import numpy
+from deap import algorithms, base, creator, gp, tools
+
+# [begin-hyperparameters]
+POPULATION_SIZE = 10
+GENERATIONS = 100
+TOURNAMENT_SIZE = 2
+CROSSOVER_RATE = 0.80
+MUTATION_RATE = 0.05
+# [end-hyperparameters]
+
+# [begin-data-parameters]
+RELATIVE_RESOURCES = "../resources/regression-data/"
+DATA_FILE = "d0.csv"
+# [end-data-parameters]
+
+
+def protected_divide(dividend: float, divisor: float) -> float:
+    """
+    The basic arithmetic operator of division. If the division basic results in a ZeroDivisionError, this function
+    returns an arbitrarily large number (999,999,999).
+
+    :param dividend: Value to divide
+    :param divisor: Value to divide by
+    :return: The quotient; the result of the division (dividend/divisor)
+    """
+    if divisor == 0:
+        return 999_999_999
```

Comment:
[math.inf](https://docs.python.org/3/library/math.html#math.inf) exists

Reply:
It will work with `inf` as it still just has a crazy bad fitness value, but it's downside is it really messes with the stats. 

![image](https://github.com/jameshughes89/cs4XX-EvolutionaryComputation/assets/1328048/c5e1d6ba-3f5d-4b91-8a9d-1e13dd40eab2)

Reply:
It may be a better approach to use an infinite value anyways, as it is more representitive of the case where you have 0 as a divisor. Then, in stats, to truncate those extreme cases to something more visual.

Alternitively, can you avoid the division by 0?

Reply:
> It may be a better approach to use an infinite value anyways, as it is more representitive of the case where you have 0 as a divisor. Then, in stats, to truncate those extreme cases to something more visual.

The trouble is, those stats come from DEAP so I don't really want to touch anything under that hood. 

> Alternitively, can you avoid the division by 0?

I _could_, but not idea. That's addressing a symptom.

Reply:
How is avoiding the division by 0 addressing a symptom? It feels more like returning infinite for a nonsensical expression is addressing the symptom -- avoiding the nonsense in the first place is not?

Reply:
The trouble is, you can't avoid it in symbolic regression unless (a) you guarantee there is no division in the expression, or (b) you have some checks on the GP such that it will never allow the divisor to be zero. The trouble with (b) is that there is the overhead and complexity of implementing it, and that would also be buried inside DEAP. People typically just assign bad fitness values to those inadmissible expressions and have them die off. 

Regardless, when I said symptom, I meant the symptom of being stuck with the DEAP stats thing (what the image is of) and having no way to address the `nan`s. However, I was wrong, I can easily do it by replacing the stat registration with the nan versions of the functions (they ignore nans) 
```python
    mstats.register("avg", numpy.nanmean)
    mstats.register("std", numpy.nanstd)
    mstats.register("min", numpy.nanmin)
    mstats.register("max", numpy.nanmax)
```
It's not perfect, because it's possible that everything in the population is nan, but it will fix the issue most of the time. 

THAT SAID, I will still leave it as the regular ones in the code but have this is a discussion in the course topic.

Reply:
Gotcha. I disagree that addresses the symptom still, rather it addresses reality. Fitness of an expression may or may not be a number. Forcing it to be one is the bandaid placed on top, and complects fitness with statistics over the fitness. Modeling reality tends to be the better option (though, I will say, it also tends to be better when you have a type system, which python fails at).